File: src/communication.cpp
#include <iostream>
#include <boost/asio.hpp>
#include <thread>
#include <chrono>
#include "message.pb.h"

using boost::asio::ip::tcp;

void send_message(boost::asio::io_context& io_context, const std::string& server, const std::string& port) {
    // Create a Person object and populate it
    Person person;
    person.set_name("Omer Faruk Okuyan");
    person.set_id(34);
    person.set_email("omerokuyan@taurob.com");

    // Serialize the object to a byte string
    std::string serialized_data;
    person.SerializeToString(&serialized_data);

    // Resolve the server address and port
    tcp::resolver resolver(io_context);
    auto endpoints = resolver.resolve(server, port);

    // Create and connect the socket
    tcp::socket socket(io_context);
    boost::asio::connect(socket, endpoints);

    // Send the serialized data
    boost::asio::write(socket, boost::asio::buffer(serialized_data));

    std::cout << "Data sent" << std::endl;
}

void handle_client(tcp::socket socket) {
    try {
        std::array<char, 1024> buf;
        boost::system::error_code error;

        size_t len = socket.read_some(boost::asio::buffer(buf), error);
        if (error == boost::asio::error::eof) {
            return; // Connection closed cleanly by peer.
        } else if (error) {
            throw boost::system::system_error(error); // Some other error.
        }

        // Deserialize the data
        Person person;
        person.ParseFromArray(buf.data(), len);

        // Print the structured data
        std::cout << "Received data:" << std::endl;
        std::cout << "Name: " << person.name() << std::endl;
        std::cout << "ID: " << person.id() << std::endl;
        std::cout << "Email: " << person.email() << std::endl;
    } catch (std::exception& e) {
        std::cerr << "Exception in thread: " << e.what() << std::endl;
    }
}

void start_receiver(boost::asio::io_context& io_context, const std::string& port) {
    tcp::acceptor acceptor(io_context, tcp::endpoint(tcp::v4(), std::stoi(port)));

    std::cout << "Receiver is running on port " << port << std::endl;

    while (true) {
        tcp::socket socket(io_context);
        acceptor.accept(socket);
        std::thread(handle_client, std::move(socket)).detach();
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <sender|receiver>" << std::endl;
        return 1;
    }

    GOOGLE_PROTOBUF_VERIFY_VERSION;

    try {
        boost::asio::io_context io_context;

        const std::string server = "127.0.0.1";
        const std::string port = "5005";

        if (std::string(argv[1]) == "sender") {
            while (true) {
                send_message(io_context, server, port);
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        } else if (std::string(argv[1]) == "receiver") {
            start_receiver(io_context, port);
        } else {
            std::cerr << "Invalid argument: " << argv[1] << std::endl;
            return 1;
        }
    } catch (std::exception& e) {
        std::cerr << "Exception: " << e.what() << std::endl;
    }

    google::protobuf::ShutdownProtobufLibrary();

    return 0;
}

File: scripts/communication.py
import asyncio
import message_pb2
import sys
from typing import Tuple

async def send_message() -> None:
    # Create a Person object and populate it
    person = message_pb2.Person()
    person.name = "John Doe"
    person.id = 1234
    person.email = "johndoe@example.com"

    # Serialize the object to a byte string
    serialized_data = person.SerializeToString()

    reader, writer = await asyncio.open_connection('127.0.0.1', 5005)

    print(f'Sending data to {writer.get_extra_info("peername")}')
    
    writer.write(serialized_data)
    await writer.drain()
    
    print('Data sent')

    writer.close()
    await writer.wait_closed()

async def periodic_send() -> None:
    while True:
        await send_message()
        await asyncio.sleep(1)

async def handle_client(reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
    data = await reader.read(1024)
    addr = writer.get_extra_info('peername')

    print(f"Connected by {addr}")

    # Deserialize the data
    person = message_pb2.Person()
    person.ParseFromString(data)

    # Print the structured data
    print("Received data:")
    print(f"Name: {person.name}")
    print(f"ID: {person.id}")
    print(f"Email: {person.email}")

    writer.close()
    await writer.wait_closed()

async def start_receiver() -> None:
    server = await asyncio.start_server(handle_client, '127.0.0.1', 5005)
    addr = server.sockets[0].getsockname()
    print(f'Serving on {addr}')

    async with server:
        await server.serve_forever()

async def main(role: str) -> None:
    if role == 'sender':
        await periodic_send()
    elif role == 'receiver':
        await start_receiver()
    else:
        print("Invalid argument: Use 'sender' or 'receiver'")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python communication.py <sender|receiver>")
    else:
        role = sys.argv[1]
        asyncio.run(main(role))
